docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,702 @@
+potentially handling this problem - see cylc-admin#128*
+
+## The New Output Syntax
+
+`foo:x` means `x` is an *expected output* of task `foo`. If a task finishes
+without completing all of its expected outputs, we will retain it in the task
+pool as an *incomplete task*.
+
+`foo?y` means `y` is an *optional output* of task `foo`. It is OK for a task to
+finish without completing its optional outputs.
+
+Abbreviated syntax for the success case:
+- `foo?` means `foo?succeed` (success is optional)
+- `foo` and `foo:` mean `foo:succeed` (success is expected)
+  - this supports Cylc 7 graphs as well as being a sensible default
+
+## Syntax Errors
+
+### succeed and fail
+
+The `succeed` and `fail` outputs are mutually exclusive opposites, therefore:
+
+- `succeed` and `fail` can't both be expected (or the task would be flagged as
+  incomplete no matter what)
+```
+# ERROR
+a
+a:fail
+```
+
+- `succeed` and `fail` can both be optional though (at a success/fail alternate
+  branching point)
+```
+# OK
+a?
+a?fail
+```
+
+- however it is OK to have one of `succeed` and `fail` expected and one
+  optional. This allows triggering off of the not-expected output at the same
+  time as the task is flagged as incomplete
+```
+# OK
+a
+a?fail => call_the_police
+```
+(Note this means we can't have *"`a` means `a:succeed` unless declared optional
+elsewhere"*.)
+
+### Custom Outputs
+
+Custom outputs must be either expected or optional, they can't be both
+```
+# ERROR
+a:x
+a?x
+```
+
+### QUESTION: submit-fail, submit, and start?
+
+It makes some sense to allow the `fail` output to be expected, but allowing
+`submit-fail` to be expected means treating a task as incomplete if it
+successfully submits. The only use case I've thought of
+is to make it somewhat easier (less cleanup before shutdown) to write
+functional tests that deliberately submit to a fake platform.
+
+Similarly, allowing `submit` and `start` to be optional equates to calling a
+task complete even though it didn't run at all, which seems equally perverse.
+
+```
+a:submit-fail  # ERROR or OK?
+a?submit  # ERROR or OK?
+a?start  # ERROR or OK?
+```
+
+## Path Branching
+
+### Concurrent Paths 
+
+Concurrent paths can trigger off of expected outputs:
+```
+a:x => b1
+a:y => b2
+b1 & b2 => c  # join
+```
+
+### Alternate Paths
+
+Alternate paths must branch off of optional outputs, and must be joined with
+OR triggers because only one side or the other will run:
+
+```
+a?x => b1
+a?y => b2
+b1 | b2 => c  # join
+```
+Note joining can trigger off of expected outputs, because the output is
+only "expected" if the task actually runs.
+
+**Failure recovery** is a common example of this:
+```
+a? => b1  # success path
+a?fail => b2  # failure path
+b1 | b2 => c  # join
+```
+
+Note that an artificial dependency can be used to **ensure that at least one
+of the alternate paths gets taken**:
+```
+a?x => b1
+a?y => b2
+b1 | b2 => c  # join
+a => c # ARTIFICIAL DEPENDENCY
```

Comment:
Do we have conventions for graphing this sort of thing? I'm finding that I have a notebook full of increasingly alchemical looking drawings...

Reply:
No, but it might be good to get more information into our visualizations, beyond graph node means task, and graph edge means a dependency of some kind.  I'm not sure I like conditionals as graph nodes though ... too much like a task called "or"?